### PR TITLE
Make `ExecutionBlockImpl.cpp` into `ExecutionBlockImpl.tpp`

### PR DESCRIPTION
**Pull request by markuspf**
### Scope & Purpose

In an attempt to improve build times, make ExecutionBlockImpl into a template header and rely on the instantiations to happen in the `createBlock` call.

In this particular case this is safe because we know that the required instantiations will happen in `createBlock` and they will also happen only once.

I also took the opportunity to address some FIXME and TODOs in ExecutionBlockImpl which were waiting for `c++17`; since we're now on `c++20` and not only do we have `if constexpr`, but also concepts.

Enterprise-Buddy: https://github.com/arangodb/enterprise/pull/1049

### CODE REVIEW

**Comment by MBkkt** [comment]
At least we needs some "benchmarks".
After this changes, and before it

For incremental compilation, and for compilation from empty build dir (and without ccache).

I think it can be really worse for incremental compilation

**Comment by MBkkt** [comment]
In general I think we still need instantiate this class only once for every unique parameter pack.
After this PR it wasn't true.

I think if you want to make building this file more parallel (do you sure it necessary, we have other files) you can separate instantiations.

But overall binary size can be bigger, because compiler cannot see every instantiations in one unit

**Comment by markuspf** [comment]
Some non-scientific numbers, on an `AMD Ryzen 9 3900X` (12 cores, 64GiB memory); Build times for Release build without tests but with IPO, `clang-14`, `lld`, no ccache. Build times for a devel version with tests are about 10 minutes longer

 * without this PR
 Compile time: `make -j 12  9015.71s user 393.56s system 1041% cpu 15:03.17 total`
File sizes
```
  8158880 bin/arangobackup
  8181424 bin/arangobench
107163504 bin/arangod
  8505560 bin/arangodump
  8207328 bin/arangoexport
  8362568 bin/arangoimport
  8405224 bin/arangorestore
 42407824 bin/arangosh
  7835840 bin/arangovpack
 ```

 * with this PR
 Compile time: `make -j 12  8433.17s user 487.30s system 1031% cpu 14:24.61 total`
File sizes
```
  8158872 bin/arangobackup
  8181416 bin/arangobench
107141984 bin/arangod
  8505552 bin/arangodump
  8207320 bin/arangoexport
  8362560 bin/arangoimport
  8405216 bin/arangorestore
 42407816 bin/arangosh
  7835832 bin/arangovpack
```

**Comment by MBkkt** [comment]
Do we really interested in release build time? Especially with lto?

**Comment by markuspf** [comment]
As I said i n the original comment, the build times for developer versions are about 10 minutes longer, alas, `RelWithDebInfo`, all tests, and without LTO

Without PR
`make -j 12  14733.70s user 595.51s system 1021% cpu 25:00.13 total`

With PR
`make -j 12  14074.27s user 558.30s system 1027% cpu 23:44.34 total`

File: arangod/Aql/ExecutionBlockImpl.tpp [hunk 1]
```diff
@@ -982,12 +900,99 @@ auto ExecutionBlockImpl<Executor>::executeSkipRowsRange(
...
-// Avoid duplicate symbols when linking: This file is directly included by
-// ExecutionBlockImplTestInstances.cpp
-#ifndef ARANGODB_INCLUDED_FROM_GTESTS
-template<>
-auto ExecutionBlockImpl<SubqueryStartExecutor>::shadowRowForwarding(
-    AqlCallStack& stack) -> ExecState {
+template<class Executor>
+auto ExecutionBlockImpl<Executor>::sideEffectShadowRowForwarding(
+    AqlCallStack& stack, SkipResult& skipResult) -> ExecState {
+  TRI_ASSERT(executorHasSideEffects<Executor>);
```
Comment (mpoeter) [inline arangod/Aql/ExecutionBlockImpl.tpp:906]: `executorHasSideEffects` should be constexpr - can we make this a static_assert?

File: arangod/Aql/ExecutionBlockImpl.tpp [hunk 2]
```diff
@@ -2408,479 +2301,48 @@ void ExecutionBlockImpl<Executor>::CallstackSplit::run(
...
-template class ::arangodb::aql::ExecutionBlockImpl<ModificationExecutor<
-    SingleRowFetcher<BlockPassthrough::Disable>, InsertModifier>>;
-template class ::arangodb::aql::ExecutionBlockImpl<ModificationExecutor<
-    SingleRowFetcher<BlockPassthrough::Disable>, RemoveModifier>>;
-template class ::arangodb::aql::ExecutionBlockImpl<ModificationExecutor<
-    SingleRowFetcher<BlockPassthrough::Disable>, UpdateReplaceModifier>>;
-template class ::arangodb::aql::ExecutionBlockImpl<ModificationExecutor<
-    SingleRowFetcher<BlockPassthrough::Disable>, UpsertModifier>>;
+// TODO: find out whether this is still needed. It is used in
+//       ScatterExecutor and in DistributeClientExecutor
```
Comment (mpoeter) [inline arangod/Aql/ExecutionBlockImpl.tpp:2305]: Is this TODO still relevant?

**Review by mpoeter — APPROVED** [review]
LGTM

**Comment by markuspf** [comment]
Green Jenkins: https://jenkins01.arangodb.biz/job/arangodb-matrix-pr-linux.x86-64/24372/

**Pull request merged**